{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "3AHsQ0DeCzY5"
      },
      "source": [
        "# HUHgle Colab Notebook\n",
        "\n",
        "<img src=\"https://drive.google.com/uc?export=view&id=1cX3z4BkUDoatQl4d20SGHL8yCkpyF6Iv\" width=\"750\" height=\"525\">\n",
        "\n",
        "This Colab notebook enables easy access to, and use of, HUHgle: a Python-based interactive tool for the visualization, design, and optimization of substrates for HUH-tag mediated covalent labeling of proteins of interest with ssDNA substrates of interest.\n",
        "\n",
        "HUHgle source code is freely available and can be accessed on Github under an open source license.\n",
        "\n",
        "**Citing this work**\n",
        "\n",
        "Any publication that discloses findings and/or figures produced by this notebook should cite the HUHgle paper.\n",
        "\n",
        "**More information**\n",
        "\n",
        "You can find more information about HUHgle in its associated paper:\n",
        "\n",
        "* [HUHgle: An Interactive Substrate Design Tool for Covalent Protein-ssDNA Labeling using HUH-tags](https://www.biorxiv.org/content/10.1101/2024.03.15.585203v1)\n",
        "\n",
        "You can find more information about HUH-tags in the following papers:\n",
        "\n",
        "* [Sequence-Directed Covalent Protein–DNA Linkages in a Single Step Using HUH-Tags](https://pubs.acs.org/doi/10.1021/jacs.7b02572)\n",
        "* [Molecular Underpinnings of ssDNA Specificity by Rep HUH-endonucleases and Implications for HUH-tag Multiplexing and Engineering](https://academic.oup.com/nar/article/49/2/1046/6067397)\n",
        "\n",
        "You can find more information about the basic science/biology of replication-initiating HUH-endonucleases (HUH-tags in their endogenous context) in the following papers:\n",
        "\n",
        "* [Breaking and joining single-stranded DNA: the HUH endonuclease superfamily](https://www.nature.com/articles/nrmicro3067)\n",
        "* [Watson-Crick Base-Pairing Requirements for ssDNA Recognition and Processing in Replication-Initiating HUH Endonucleases](https://journals.asm.org/doi/10.1128/mbio.02587-22)\n",
        "\n",
        "**Anatomy of an HUH-tag Reaction**\n",
        "\n",
        "HUH-tags are versatile fusion proteins that mediate sequence-specific protein-ssDNA covalent linkage in a simple, efficient, and cost-effective manner. These fusion partners have emerged as an ideal tool to link proteins of interest to ssDNA sequences of interest because of their minimal requirements for reaction and their long-lived protein-ssDNA bioconjugates.\n",
        "\n",
        "HUH-tags use an eponymous metal coordinating motif most often comprised of two histidine (H) separated by a bulky hydrophobic residue (U) to coordinate a divalent cation -- either magnesium or manganese -- to their active site which the enzyme uses to polarize the phosphate backbone of a targeted ssDNA substrate, enabling the nuclease's catalytic tyrosine to cleave the nucleic acid and then covalently bind to its newly exposed 5' end.\n",
        "\n",
        "See the image below for a graphical depiction of this process.\n",
        "\n",
        "<img src=\"https://drive.google.com/uc?export=view&id=1uqfqD7zAZFjoo23S_zvNGatXi5ujdoq6\" width=\"975\" height=\"375\">\n",
        "\n",
        "**How to interpret HUHgle plots**\n",
        "\n",
        "HUHgle plots color each of the four nucleobases differently in order to enhance interpretation -- As, Ts, Gs, and Cs are yellow, orange, green, and red, respectively. Importantly, if HUHgle identifies a sequence with which the user-selected HUH-tag interacts it recolors it to a monochromatic blue gradient indicative of how good of a substrate that sequence is for the HUH-tag. The darker blue a sequence is recolored, the more effective the selected HUH-tag is at interacting with that substrate. See the 'Substrate Score' legend on the right side of the plot. Substrates with scores of 0.8 or higher have their text recolored white for clarity. All HUH-tag substrates are comprised of seven bases upstream of the cleavage site (positions -7 through -1) and two bases downstream of the cleavage site (+1 and +2) -- the covalent bond is formed to the base in position +1. HUHgle identifies cleavage sites by highlighting them with a black vertical line. The base immediately to the right of one of these lines is the +1 position.\n",
        "\n",
        "In instances where there are two or more sequences that overlap with one another, HUHgle highlights where each contiguous sequence is in the greater sequence context by underlining it with a black horizontal line. When two black horizontal lines partially overlap, as they do across sites two and three in the image below, it indicates the presence of overlapping sequences. The cleavage sites on overlapping sequences are still appropriately indicated. The color indicative of the 'Substrate Score' metric is recolored from the first overlapping base onwards to indicate the efficiency of the second substrate while retaining the color of the underlying first substrate in all positions preceding the overlap.\n",
        "\n",
        "<img src=\"https://drive.google.com/uc?export=view&id=1JeWEp5MJazFL_vL6MQU4LDpu6ywf1SnB\" width=\"750\" height=\"420\">\n",
        "\n",
        "\n",
        "**Contact**\n",
        "\n",
        "If you have any questions or comments, please contact Adam Smiley (adamtreysmiley@gmail.com) and Wendy Gordon (wrgordon@umn.edu)\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "pQYdk6Nxufxk"
      },
      "source": [
        "# HUHgle Search Function\n",
        "\n",
        "The following cells, labeled 1 - 4, comprise the core functionality of HUHgle. Select your experiment type (i.e., single, two-, or three-way labeling), enter your substrate sequence(s), select your HUH-tag(s), and then search & plot!\n",
        "\n",
        "*   Progress through the cells in a linear manner (i.e., run through the program cell-by-cell) for the best experience.\n",
        "*   Carefully read the instructions in each cell and respond to prompts/dropdown menus as they appear.\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "nvaO_CPb6cUo"
      },
      "outputs": [],
      "source": [
        "#@title 1: Install Required Modules, Load Data, and Declare Functions\n",
        "#@markdown 1. Run this cell.\n",
        "#@markdown 2. Wait for the download to complete.\n",
        "#@markdown  * This should only take a handful of seconds.\n",
        "#@markdown 3. Proceed to the next cell once the terminal clears and indicates that installation is complete.\n",
        "\n",
        "import matplotlib.pyplot as plt\n",
        "import matplotlib.patches as patches\n",
        "from matplotlib.cm import Blues\n",
        "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
        "from matplotlib.colorbar import ColorbarBase\n",
        "from matplotlib.colors import Normalize\n",
        "import matplotlib.transforms as transforms\n",
        "import pandas as pd\n",
        "import math\n",
        "import requests\n",
        "from google.colab import files\n",
        "from IPython.display import display, clear_output\n",
        "import zipfile\n",
        "from itertools import combinations\n",
        "from statistics import mean\n",
        "from google.colab import output\n",
        "from matplotlib.font_manager import FontProperties\n",
        "output.enable_custom_widget_manager()\n",
        "\n",
        "# Install required stuff that doesn't come with Colab\n",
        "!pip install ipympl\n",
        "!pip install tqdm\n",
        "!pip install ViennaRNA\n",
        "\n",
        "##\n",
        "# Load and Prepare Data\n",
        "# Direct download link to the file stored on Gordon Lab Google Drive\n",
        "file_url = \"https://drive.google.com/uc?export=download&id=15WmYH0sLUhmwmsiqjjYcQH-EWqe2bjlc\"\n",
        "\n",
        "# Download the file using download link\n",
        "r = requests.get(file_url, allow_redirects=True)\n",
        "open('SubstrateScores.csv', 'wb').write(r.content)\n",
        "\n",
        "# Load the dataset\n",
        "data = pd.read_csv('SubstrateScores.csv')\n",
        "\n",
        "# Drop the less useful information -- CpCDV and MSMV are too promiscuous to be\n",
        "# useful, and WDV(106F) is a catalytically inactive control\n",
        "data.drop(columns=['CpCDV', 'MSMV', 'WDV(Y106F)'], inplace=True)\n",
        "\n",
        "# Add 'AC' to the end of each k-mer in the dataset -- interaction sites conclude\n",
        "# with an 'AC' dinucleotide\n",
        "data['kmers'] = data['kmers'].apply(lambda x: x + 'AC')\n",
        "\n",
        "# Function to create a dictionary for a given HUH-tag filtering out sequences with\n",
        "# substrate scores less than 0.05\n",
        "def create_dict_from_column(dataframe, column_name):\n",
        "    return {k: v for k, v in zip(dataframe['kmers'], dataframe[column_name]) if v > 0.05}\n",
        "\n",
        "# Create a dictionary for each column (except the first two, which are labels)\n",
        "dictionaries = {col: create_dict_from_column(data, col) for col in data.columns[2:]}\n",
        "\n",
        "# Global variables used in future functions\n",
        "sequences = []\n",
        "current_sequence_index = 0\n",
        "nucs = list(dictionaries.keys())\n",
        "selected_option = None\n",
        "\n",
        "###\n",
        "# Declare Functions\n",
        "\n",
        "# Sequence filtering functions\n",
        "def contains_non_ATGC_characters(sequence_list):\n",
        "    for sequence in sequence_list:\n",
        "        for char in sequence:\n",
        "            if char not in ['A', 'T', 'G', 'C']:\n",
        "                return True\n",
        "    return False\n",
        "\n",
        "def check_long_short_strings(string_list):\n",
        "    for string in string_list:\n",
        "        if len(string) > 300:\n",
        "            return True\n",
        "        elif len(string) < 9:\n",
        "            return True\n",
        "    return False\n",
        "\n",
        "def has_duplicates(string_list):\n",
        "    seen_strings = set()\n",
        "\n",
        "    for string in string_list:\n",
        "        if string in seen_strings:\n",
        "            return True\n",
        "        seen_strings.add(string)\n",
        "\n",
        "    return False\n",
        "\n",
        "# Function to recommend three HUH-tags based on the substrate queried in the context of single-labeling\n",
        "def recommend_huh(substrate, data):\n",
        "\n",
        "    # Search the substrate for scored sequences and calculate the total percent reduction for each nuclease\n",
        "    total_reduction = {col: 0 for col in data.columns if col not in ['k-mer #', 'kmers']}\n",
        "    for _, row in data.iterrows():\n",
        "        kmer = row['kmers']\n",
        "        count = substrate.count(kmer)\n",
        "        for nuclease in total_reduction.keys():\n",
        "            total_reduction[nuclease] += (row[nuclease] * count)\n",
        "\n",
        "    # Check if all score values are zero\n",
        "    if all(value == 0 for value in total_reduction.values()):\n",
        "        # If all values are zero, return all nuclease options\n",
        "        print(\"\\nNo preferred HUH-tag -- no predicted cleavage across all candidates -- select any of the eight options.\")\n",
        "        return list(total_reduction.keys())\n",
        "    elif all(abs(value - mean(total_reduction.values())) / mean(total_reduction.values()) <= 0.05 for value in total_reduction.values()):\n",
        "        # Check if all values are within 5% of their mean\n",
        "        print(\"\\nNo preferred HUH-tag -- nearly equivalent predicted cleavage across all candidates -- select any of the eight options.\")\n",
        "        return list(total_reduction.keys())\n",
        "    else:\n",
        "        # Otherwise, return the top three keys\n",
        "        print(\"\\nHUHgle has recommended its top three HUH-tags based on lowest interaction with the specified substrate\")\n",
        "        top_three_keys = sorted(total_reduction, key=total_reduction.get, reverse=False)[:3]\n",
        "        return top_three_keys\n",
        "\n",
        "# Logic and functions to score and recommend HUH-tag pairs, triplets, and quadruplets\n",
        "\n",
        "# Define thresholds\n",
        "\n",
        "# Scored need to be higher than 0.5\n",
        "upper_threshold = 0.50\n",
        "# Scored can't be lower than 0.1\n",
        "lower_threshold = 0.10\n",
        "\n",
        "# List of HUH-tags\n",
        "nucleases = data.columns[2:]\n",
        "\n",
        "# Function to calculate orthogonality score\n",
        "def calculate_orthogonality_score(orthogonal_kmers, nucleases):\n",
        "    scores = {}\n",
        "    for key, value in orthogonal_kmers.items():\n",
        "        if not value.empty:\n",
        "            score = 1\n",
        "            count_diff = 0\n",
        "            for nuclease in key:\n",
        "                if nuclease in value.columns:\n",
        "                    count = sum(value[nuclease] >= upper_threshold)\n",
        "                    score *= count\n",
        "                    count_diff += count\n",
        "            score -= abs(count_diff)\n",
        "            scores[key] = score\n",
        "        else:\n",
        "            scores[key] = 0\n",
        "    return scores\n",
        "\n",
        "# Function to find orthogonal substrates for pairs and triplets\n",
        "def find_orthogonal_kmers(data, nucleases, num_nucleases):\n",
        "    orthogonal_kmers_dict = {}\n",
        "    for nuclease_combination in combinations(nucleases, num_nucleases):\n",
        "        key = frozenset(nuclease_combination)\n",
        "        orthogonal_kmers = pd.DataFrame()\n",
        "        for nuclease in nuclease_combination:\n",
        "            conditions = [(data[nuclease] >= upper_threshold) if nuclease == n else (data[n] <= lower_threshold) for n in nuclease_combination]\n",
        "            selected_kmers = data[conditions[0]]\n",
        "            for condition in conditions[1:]:\n",
        "                selected_kmers = selected_kmers[condition]\n",
        "            if not selected_kmers.empty:\n",
        "                orthogonal_kmers = pd.concat([orthogonal_kmers, selected_kmers]).drop_duplicates()\n",
        "\n",
        "        orthogonal_kmers_dict[key] = orthogonal_kmers[['kmers'] + list(nuclease_combination)]\n",
        "\n",
        "    return orthogonal_kmers_dict\n",
        "\n",
        "# Finding orthogonal k-mers for pairs, triplets, and quadruplets\n",
        "orthogonal_kmers_pairs = find_orthogonal_kmers(data, nucleases, 2)\n",
        "orthogonal_kmers_triplets = find_orthogonal_kmers(data, nucleases, 3)\n",
        "orthogonal_kmers_quadruplets = find_orthogonal_kmers(data, nucleases, 4)\n",
        "\n",
        "# Scoring orthogonal sequences for pairs and triplets\n",
        "orthogonality_scores_pairs = calculate_orthogonality_score(orthogonal_kmers_pairs, nucleases)\n",
        "orthogonality_scores_triplets = calculate_orthogonality_score(orthogonal_kmers_triplets, nucleases)\n",
        "orthogonality_scores_quadruplets = calculate_orthogonality_score(orthogonal_kmers_quadruplets, nucleases)\n",
        "\n",
        "# Function to select top scoring sequences from the identified orthogonal sequences\n",
        "def select_top_scoring(orthogonality_scores, orthogonal_kmers_dict, top_n):\n",
        "    top_scores = sorted(orthogonality_scores.items(), key=lambda x: x[1], reverse=True)[:top_n]\n",
        "    return {key: orthogonal_kmers_dict[key] for key, _ in top_scores}\n",
        "\n",
        "# Selecting top scoring pairs and triplets\n",
        "top_scoring_pairs = select_top_scoring(orthogonality_scores_pairs, orthogonal_kmers_pairs, 10)\n",
        "top_scoring_triplets = select_top_scoring(orthogonality_scores_triplets, orthogonal_kmers_triplets, 10)\n",
        "top_scoring_quadruplets = select_top_scoring(orthogonality_scores_quadruplets, orthogonal_kmers_quadruplets, 3)  # Not in the manuscript or HUHgle, but accessible here\n",
        "\n",
        "# Function to recommend three HUH-tags based on the substrate queried in the context of single-labeling\n",
        "def recommend_huh_pair(substrates, data, available_pairs):\n",
        "    # Initialize total score dictionary for individuals\n",
        "    total_score_individuals = {col: 0 for col in data.columns if col not in ['k-mer #', 'kmers']}\n",
        "\n",
        "    # Calculate total score for each individual nuclease\n",
        "    for substrate in substrates:\n",
        "        for _, row in data.iterrows():\n",
        "            kmer = row['kmers']\n",
        "            count = substrate.count(kmer)\n",
        "            for nuclease in total_score_individuals.keys():\n",
        "                total_score_individuals[nuclease] += (row[nuclease] * count)\n",
        "\n",
        "    # Calculate combined scores for each pair in available pairs\n",
        "    pair_scores = {}\n",
        "    for pair in available_pairs:\n",
        "        nuclease1, nuclease2 = pair.split(', ')\n",
        "        combined_score = total_score_individuals.get(nuclease1, 0) + total_score_individuals.get(nuclease2, 0)\n",
        "        pair_scores[pair] = combined_score\n",
        "\n",
        "    # Check if all scores are zero\n",
        "    if all(score == 0.0 for score in pair_scores.values()):\n",
        "        print(\"\\nNo preferred HUH-tag pair -- no predicted cleavage across all candidate pairs -- select any of the options.\")\n",
        "        return available_pairs\n",
        "\n",
        "    # Check if all scores are within 5% of each other\n",
        "    if all(abs(score - mean(pair_scores.values())) / mean(pair_scores.values()) <= 0.05 for score in pair_scores.values()):\n",
        "        print(\"\\nNo preferred HUH-tag pair -- nearly equivalent predicted cleavage across all candidate pairs -- select any of the options.\")\n",
        "        return available_pairs\n",
        "\n",
        "    # Sort pairs by score and select top three\n",
        "    top_three_pairs = sorted(pair_scores, key=pair_scores.get)[:3]\n",
        "    print(\"\\nHUHgle has recommended its top three HUH-tags pairs based on lowest interaction with the specified substrate(s)\")\n",
        "    return top_three_pairs\n",
        "\n",
        "# Function to recommend three HUH-tags based on the substrate queried in the context of single-labeling\n",
        "def recommend_huh_triplet(substrates, data, available_triplets):\n",
        "    # Initialize total reduction dictionary for individuals\n",
        "    total_score_individuals = {col: 0 for col in data.columns if col not in ['k-mer #', 'kmers']}\n",
        "\n",
        "    # Calculate total reduction for each individual nuclease\n",
        "    for substrate in substrates:\n",
        "        for _, row in data.iterrows():\n",
        "            kmer = row['kmers']\n",
        "            count = substrate.count(kmer)\n",
        "            for nuclease in total_score_individuals.keys():\n",
        "                total_score_individuals[nuclease] += (row[nuclease] * count)\n",
        "\n",
        "    # Calculate combined scores for each triplet in available triplets\n",
        "    triplet_scores = {}\n",
        "    for triplet in available_triplets:\n",
        "        nucleases = triplet.split(', ')\n",
        "        combined_score = sum(total_score_individuals.get(nuclease, 0) for nuclease in nucleases)\n",
        "        triplet_scores[triplet] = combined_score\n",
        "\n",
        "    # Check if all scores are zero\n",
        "    if all(score == 0.0 for score in triplet_scores.values()):\n",
        "        print(\"\\nNo preferred HUH-tag triplet -- no predicted cleavage across all candidate triplets -- select any of the options.\")\n",
        "        return available_triplets\n",
        "\n",
        "    # Check if all scores are within 5% of each other\n",
        "    if all(abs(score - mean(triplet_scores.values())) / mean(triplet_scores.values()) <= 0.05 for score in triplet_scores.values()):\n",
        "        print(\"\\nNo preferred HUH-tag triplet -- nearly equivalent predicted cleavage across all candidate triplets -- select any of the options.\")\n",
        "        return available_triplets\n",
        "\n",
        "    # Sort triplets by score and select top three\n",
        "    top_three_triplets = sorted(triplet_scores, key=triplet_scores.get)[:3]\n",
        "\n",
        "    print(\"\\nHUHgle has recommended its top three HUH-tags triplets based on lowest interaction with the specified substrate(s)\")\n",
        "    return top_three_triplets\n",
        "\n",
        "# HUHgle plot functions -- primary and accessory functions\n",
        "\n",
        "# Primary functions for interactive plot\n",
        "def recolor_rects_for_sequences(rects, texts, colors, sequence_intensity_dict, ax):\n",
        "    for rect, text in zip(rects, texts):\n",
        "        rect.set_facecolor(colors[text.get_text()])\n",
        "        text.set_color('black')\n",
        "\n",
        "    sequence_counter = 1\n",
        "\n",
        "    motif_indices_dict = {}\n",
        "\n",
        "    for child in ax.get_children():\n",
        "        if isinstance(child, patches.Rectangle) and (child.get_facecolor() == (0, 0, 0, 1) or child.get_label() == 'underline'):\n",
        "            child.remove()\n",
        "\n",
        "    # Color sequences based on intensity using a blue colormap and white text\n",
        "    for idx, text in enumerate(texts):\n",
        "        for sequence, intensity in sequence_intensity_dict.items():\n",
        "            sequence_length = len(sequence)\n",
        "            if idx + sequence_length <= len(texts):\n",
        "                current_sequence = ''.join(inner_text.get_text() for inner_text in texts[idx:idx + sequence_length])\n",
        "                if current_sequence == sequence:\n",
        "                    # Update the motif indices dictionary\n",
        "                    sequence_label = f'Seq{sequence_counter}'\n",
        "                    sequence_counter += 1\n",
        "                    motif_indices_dict[sequence_label] = [idx, idx + sequence_length - 1]\n",
        "\n",
        "                    color = Blues(intensity)  # Get the color from the Blues colormap\n",
        "\n",
        "                    for j, rect in enumerate(rects[idx:idx + sequence_length]):\n",
        "                        rect.set_facecolor(color)\n",
        "                        # Change text color to white if intensity is 0.8 or higher\n",
        "                        if intensity >= 0.8:\n",
        "                            texts[idx + j].set_color('white')\n",
        "                        else:\n",
        "                            texts[idx + j].set_color('black')\n",
        "\n",
        "                        # Draw a black bar to indicate cleavage site\n",
        "                        if j == sequence_length - 2:\n",
        "                            left_edge = rect.get_x()\n",
        "                            bottom_edge = rect.get_y()\n",
        "                            height = rect.get_height()\n",
        "                            black_bar = patches.Rectangle((left_edge, bottom_edge), -0.05, height, color='black')\n",
        "                            ax.add_patch(black_bar)\n",
        "                    break\n",
        "\n",
        "    def check_overlap(seq1, seq2):\n",
        "        # Overlap occurs if the end index of seq1 is at least the start index of seq2\n",
        "        return seq1[1] >= seq2[0]\n",
        "\n",
        "    def add_spacing_values(motif_indices_dict):\n",
        "        sequence_labels = list(motif_indices_dict.keys())\n",
        "\n",
        "        for label in sequence_labels:\n",
        "            motif_indices_dict[label].append(None)\n",
        "\n",
        "        current_spacing = 0.1\n",
        "        for i in range(len(sequence_labels) - 1):\n",
        "            current_label = sequence_labels[i]\n",
        "            next_label = sequence_labels[i + 1]\n",
        "\n",
        "            if check_overlap(motif_indices_dict[current_label], motif_indices_dict[next_label]):\n",
        "                if motif_indices_dict[current_label][2] is None:\n",
        "                    motif_indices_dict[current_label][2] = current_spacing\n",
        "\n",
        "                current_spacing = current_spacing + 0.1\n",
        "                if current_spacing > 0.4:\n",
        "                    current_spacing = 0.1\n",
        "                motif_indices_dict[next_label][2] = current_spacing\n",
        "            else:\n",
        "                current_spacing = 0.1\n",
        "\n",
        "    add_spacing_values(motif_indices_dict)\n",
        "\n",
        "    nts_per_row = 15  # Number of nucleotides per row - 15 looks the nicest in my opinion\n",
        "\n",
        "    for sequence_label, seq_info in motif_indices_dict.items():\n",
        "        start_idx, end_idx, underline_spacing = seq_info\n",
        "\n",
        "        if underline_spacing is None:\n",
        "            continue\n",
        "\n",
        "        current_idx = start_idx\n",
        "        while current_idx <= end_idx:\n",
        "            segment_start = current_idx\n",
        "            segment_end = min(end_idx, (current_idx // nts_per_row + 1) * nts_per_row - 1)\n",
        "\n",
        "            box_width = rects[segment_start].get_width()\n",
        "            adjustment = 0.05 * box_width\n",
        "\n",
        "            start_left_edge = rects[segment_start].get_x() + adjustment\n",
        "            end_right_edge = rects[segment_end].get_x() + rects[segment_end].get_width() - adjustment\n",
        "            underline_length = end_right_edge - start_left_edge\n",
        "\n",
        "            bottom_edge = rects[segment_start].get_y()\n",
        "            underline_height = 0.03\n",
        "            underline = patches.Rectangle((start_left_edge, bottom_edge + rects[segment_start].get_height() + underline_spacing), underline_length, underline_height, color='black', label='underline')\n",
        "            ax.add_patch(underline)\n",
        "\n",
        "            current_idx = segment_end + 1\n",
        "\n",
        "def onclick(event, rects, texts, ax, colors, sequence_intensity_dict, modified_sequence):\n",
        "    base_order = ['A', 'T', 'G', 'C']\n",
        "    for i, (rect, text) in enumerate(zip(rects, texts)):\n",
        "        if rect.contains(event)[0]:\n",
        "            current_base = text.get_text()\n",
        "            next_base = base_order[(base_order.index(current_base) + 1) % len(base_order)]\n",
        "            text.set_text(next_base)\n",
        "\n",
        "            # Recolor all rectangles to update sequences\n",
        "            recolor_rects_for_sequences(rects, texts, colors, sequence_intensity_dict, ax)\n",
        "\n",
        "            # Update the modified sequence\n",
        "            modified_sequence[i] = next_base\n",
        "\n",
        "            ax.figure.canvas.draw()\n",
        "            break\n",
        "\n",
        "def plot_dna_sequence(sequence_list, modified_sequence, nts_per_row=15, show_legend=False):\n",
        "    global current_sequence_index, virus_name, temp\n",
        "    sequence_str = sequence_list[current_sequence_index]\n",
        "\n",
        "    row_height = 1.6\n",
        "    fig_width = (nts_per_row + 3)\n",
        "    fig_height = row_height * (math.ceil(len(sequence_str) / nts_per_row) + 1)\n",
        "    fig, ax = plt.subplots(figsize=(fig_width, fig_height))\n",
        "\n",
        "    colors = {\n",
        "        'A': '#F2C14E',\n",
        "        'T': '#F78154',\n",
        "        'G': '#5FAD56',\n",
        "        'C': '#B4436C'\n",
        "    }\n",
        "    number_bg_color = '#D3D3D3'\n",
        "\n",
        "    rects = []\n",
        "    texts = []\n",
        "\n",
        "    x = 1.5\n",
        "    y = 0\n",
        "    width = 1\n",
        "    height = 1\n",
        "    current_row = 1\n",
        "\n",
        "    font_size = 24\n",
        "\n",
        "    for idx, base in enumerate(sequence_str):\n",
        "        color = colors[base]\n",
        "\n",
        "        rect = patches.Rectangle((x, y), width, height, facecolor=color)\n",
        "        ax.add_patch(rect)\n",
        "        text = ax.text(x + 0.5, y + 0.5, base, ha='center', va='center', color='black', weight='bold', fontsize=font_size)\n",
        "\n",
        "        rects.append(rect)\n",
        "        texts.append(text)\n",
        "\n",
        "        if (idx + 1) % nts_per_row == 0 or (idx + 1) == len(sequence_str):\n",
        "            start_number = (current_row - 1) * nts_per_row + 1\n",
        "            ax.add_patch(patches.Rectangle((0, y), 1.5, height, facecolor=number_bg_color))\n",
        "            ax.add_patch(patches.Rectangle((x + 1, y), 1.5, height, facecolor=number_bg_color))\n",
        "            ax.text(0.75, y + 0.5, str(start_number), va='center', ha='center', color='black', fontsize=18, weight='bold')\n",
        "            ax.text(x + 1.75, y + 0.5, str(idx + 1), va='center', ha='center', color='black', fontsize=18, weight='bold')\n",
        "\n",
        "            y += row_height\n",
        "            x = 1.5\n",
        "            current_row += 1\n",
        "        else:\n",
        "            x += width\n",
        "\n",
        "    # Initially recolor all rectangles to update sequences\n",
        "    recolor_rects_for_sequences(rects, texts, colors, sequence_intensity_dict, ax)\n",
        "\n",
        "    ax.set_xlim(0, nts_per_row + 3)\n",
        "    num_rows = len(sequence_str) // nts_per_row\n",
        "    if len(sequence_str) % nts_per_row > 0:\n",
        "        num_rows += 1\n",
        "\n",
        "    ax.set_ylim(0, row_height * num_rows)\n",
        "    ax.axis('off')\n",
        "    plt.gca().invert_yaxis()\n",
        "\n",
        "    legend_font = FontProperties(weight='bold', size=24)\n",
        "\n",
        "    if show_legend:\n",
        "        legend_elements = [patches.Patch(facecolor=color, edgecolor='none', label=base) for base, color in colors.items()]\n",
        "        ax.legend(handles=legend_elements, loc='upper center', bbox_to_anchor=(0.5, 0.0), ncol=len(colors), edgecolor='white', prop=legend_font)\n",
        "\n",
        "    # Find the virus name from the sequence_intensity_dict\n",
        "    virus_name = [key for key, value in dictionaries.items() if value == sequence_intensity_dict][0]\n",
        "\n",
        "    # Set the title for the plot using the virus name\n",
        "    if len(sequence_list) == 1:\n",
        "        ax.set_title(f'{virus_name} Cleavage Profile', fontsize=36, weight='bold', pad=30)\n",
        "    else:\n",
        "        ax.set_title(f'{virus_name} Cleavage Profile - Substrate {current_sequence_index + 1}', fontsize=36, weight='bold', pad=30)\n",
        "\n",
        "    divider = make_axes_locatable(ax)\n",
        "    cax = divider.append_axes(\"right\", size=\"2.5%\", pad=0.5)\n",
        "    norm = Normalize(vmin=0, vmax=1)\n",
        "    cb = ColorbarBase(cax, cmap=Blues, norm=norm, orientation='vertical')\n",
        "\n",
        "    if len(sequence_str) <= 15:\n",
        "        cb.set_label('Substrate\\nScore', fontsize=28, weight='bold')\n",
        "    else:\n",
        "        cb.set_label('\\nSubstrate Score', fontsize=28, weight='bold')\n",
        "\n",
        "    # Set tick labels to bold\n",
        "    for label in cb.ax.yaxis.get_ticklabels():\n",
        "        label.set_weight('bold')\n",
        "        label.set_fontsize(18)\n",
        "\n",
        "    # Define modified_sequence to track changes\n",
        "    modified_sequence = list(sequence_str)\n",
        "    temp = modified_sequence\n",
        "\n",
        "    # Connect the onclick event and pass modified_sequence\n",
        "    fig.canvas.mpl_connect('button_press_event', lambda event: onclick(event, rects, texts, ax, colors, sequence_intensity_dict, modified_sequence))\n",
        "\n",
        "    plt.tight_layout()\n",
        "    plt.gcf().set_dpi(75)\n",
        "    plt.show()\n",
        "\n",
        "# Accessory functions for cleavage site and reaction product information\n",
        "\n",
        "def calculate_dna_molecular_weight(dna_sequence):\n",
        "    # Average molecular weights of the nucleotides in g/mol\n",
        "    weights = {\n",
        "        'A': 313.21,\n",
        "        'T': 304.2,\n",
        "        'C': 289.18,\n",
        "        'G': 329.21\n",
        "    }\n",
        "\n",
        "    # Calculate the total weight of the nucleotides\n",
        "    nucleotide_weight = (sum(weights[nucleotide] for nucleotide in dna_sequence))\n",
        "\n",
        "    # Convert g/mol to kDa (1 g/mol = 1 Da, and 1000 Da = 1 kDa)\n",
        "    molecular_weight_kDa = round(nucleotide_weight / 1000, 2)\n",
        "\n",
        "    return molecular_weight_kDa\n",
        "\n",
        "def find_primary_products(sequence, sequences_of_interest):\n",
        "# Primary products are those produces from an HUH-tag interacting with a substrate\n",
        "    results = []\n",
        "    for seq in sequences_of_interest:\n",
        "        index = sequence.find(seq)\n",
        "        while index != -1:\n",
        "            # Append the part of the sequence from the match to the end\n",
        "            results.append(sequence[index+7:])\n",
        "            # Find next occurrence by searching in the remaining sequence\n",
        "            index = sequence.find(seq, index + 1)\n",
        "    return results\n",
        "\n",
        "def find_secondary_products(results, sequences_of_interest):\n",
        "# Secondary products are those produces from an HUH-tag interacting with a substrate\n",
        "# that is already covalently bound to an HUH-tag, reducing the size of the initial conjugate\n",
        "# this is rare to see in a bioconjugation reaction with 1-to-1 enzyme-substrate ratios\n",
        "    extracted_sequences = []\n",
        "    for result in results:\n",
        "        for seq in sequences_of_interest:\n",
        "            index = result.find(seq)\n",
        "            if index != -1:\n",
        "                # Extract the part of the substring from the beginning up to index 6 of the sequence of interest\n",
        "                extracted_sequences.append(result[:index+7])\n",
        "    return extracted_sequences\n",
        "\n",
        "clear_output()\n",
        "print(\"Installation complete - proceed to the next cell.\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "-nCj7tlkEHeF"
      },
      "outputs": [],
      "source": [
        "#@title 2: Select Experiment Type and Input ssDNA Substrate Sequence(s) to Query\n",
        "#@markdown 1. Run this cell. (More specificity in what will happen when you click run)\n",
        "#@markdown 2. Select experiment type from the dropdown menu (Single, Two-, or Three-Way Labeling).\n",
        "#@markdown 3. Enter ssDNA substrate sequence(s) to search (sequences must be between 9 and 300 nt in length).\n",
        "#@markdown  * HUHgle can include an optional base color key at the bottom of the plots it produces.\n",
        "#@markdown  * You can either add your own HUH-motif at the 5' end of your ssDNA substrate(s), or HUHgle can append one to your substrate(s) based on your selected/recommended HUH-tag(s).\n",
        "#@markdown  * It is **highly recommended that you use HUHgle to append HUH-motifs to your substrates in orthogonal labeling reactions** because there are so few combinations that enable non-cross reactive, unique cleavage events.\n",
        "#@markdown  * See the schematic below for a graphical representation of the different experiment options:\n",
        "\n",
        "#@markdown <img src=\"https://drive.google.com/uc?export=view&id=1H-idaK-VOwBF3BU5JvjPxU0Bi-GPMa2B\" width=\"1665\" height=\"300\">\n",
        "\n",
        "#@markdown 4. Hit the \"Submit Sequence(s)\" button and move on to the next cell.\n",
        "\n",
        "###\n",
        "\n",
        "if 'data' not in locals() and 'data' not in globals():\n",
        "    raise NameError(\"PLEASE RUN CELL 1 BEFORE ATTEMPTING TO RUN OTHER CELLS!\")\n",
        "\n",
        "###\n",
        "\n",
        "import ipywidgets as widgets\n",
        "from IPython.display import display, clear_output\n",
        "\n",
        "print(\"Welcome to HUHgle Colab!\\n\\n1. Select 'Experiment Type' from the dropdown menu below.\\n\\n    - See the reaction schematics above for more details on possible experiment options.\\n\\n    - For two- and three-way labeling, enter the desired number of substrate(s) and leave undesired sequence fields blank.\\n\\n2. Enter the ssDNA substrate sequence(s) to search ⬇️\\n\\n3. Hit the 'Submit Sequence(s)' button and proceed to the next cell.\\n\")\n",
        "\n",
        "# HTML Widget for the Dropdown Label\n",
        "dropdown_label = widgets.HTML(\n",
        "    value=\"<b>Experiment Type:</b>\",\n",
        "    layout=widgets.Layout(margin='0 0 5px 0')\n",
        ")\n",
        "\n",
        "# Dropdown for Experiment Type\n",
        "experiment_type_dropdown = widgets.Dropdown(\n",
        "    options=[('', 0), ('Single-Labeling', 1), ('Two-Way Labeling', 2), ('Three-Way Labeling', 3)],\n",
        "    value=0,\n",
        "    description='',\n",
        ")\n",
        "\n",
        "# Text Input Widgets for Labeling\n",
        "text_inputs = [widgets.Text(description=f'Sequence {i+1}:') for i in range(3)]\n",
        "\n",
        "# Container for Text Inputs and Checkbox\n",
        "input_container = widgets.VBox()\n",
        "\n",
        "# Global variable to store the legend display state\n",
        "global_legend = False\n",
        "\n",
        "# Function to update the global variable when the checkbox is toggled\n",
        "def update_legend_state(change):\n",
        "    global global_legend\n",
        "    global_legend = change.new\n",
        "\n",
        "# Legend Toggle Checkbox -- New Addition\n",
        "legend_toggle = widgets.Checkbox(\n",
        "    value=False,\n",
        "    disabled=False,\n",
        "    indent=False\n",
        ")\n",
        "\n",
        "# Attach the observer to the legend toggle checkbox\n",
        "legend_toggle.observe(update_legend_state, names='value')\n",
        "\n",
        "# Label for the legend toggle checkbox\n",
        "legend_description = widgets.Label(\n",
        "    value=\"Click this box to include a base color key on the HUHgle plot.\",\n",
        "    layout=widgets.Layout(margin='0 0 0 -285px')  # Adjust margin as needed\n",
        ")\n",
        "\n",
        "# Arrange the Checkbox and Label in a Horizontal Box\n",
        "legend_with_description = widgets.HBox([legend_toggle, legend_description])\n",
        "\n",
        "# Global variable for the checkbox state\n",
        "checkbox_state = False\n",
        "\n",
        "# Checkbox Widget\n",
        "checkbox_widget = widgets.Checkbox(\n",
        "    value=False,\n",
        "    disabled=False,\n",
        "    indent=False\n",
        ")\n",
        "\n",
        "# Label Widget for the checkbox\n",
        "checkbox_description = widgets.Label(\n",
        "    value=\"Click this box if you want HUHgle to append an HUH-motif to the 5' end of your substrate.\",\n",
        "    layout=widgets.Layout(margin='0 0 0 -285px')\n",
        ")\n",
        "\n",
        "# Arrange the Checkbox and Label in a Horizontal Box\n",
        "checkbox_with_description = widgets.HBox([checkbox_widget, checkbox_description])\n",
        "\n",
        "# Global variable for multi-labeling\n",
        "two_labeling = False\n",
        "three_labeling = False\n",
        "\n",
        "def update_checkbox_state(change):\n",
        "    global checkbox_state\n",
        "    checkbox_state = change.new\n",
        "\n",
        "# Observe checkbox for changes and update checkbox_state accordingly\n",
        "checkbox_widget.observe(update_checkbox_state, 'value')\n",
        "\n",
        "submit_button = widgets.Button(description=\"Submit Sequence(s)\", layout=widgets.Layout(margin='10px 0 0 0'))\n",
        "\n",
        "# Update Input Fields Based on Experiment Type -- Updated Section\n",
        "def update_input_fields(*args):\n",
        "    global two_labeling, three_labeling\n",
        "\n",
        "    number_of_inputs = experiment_type_dropdown.value\n",
        "    elements_to_display = [legend_with_description]\n",
        "    if number_of_inputs > 0:\n",
        "        elements_to_display.append(checkbox_with_description)\n",
        "    elements_to_display.extend(text_inputs[:number_of_inputs])\n",
        "\n",
        "    # Button Widget inside this function\n",
        "    if number_of_inputs > 0:\n",
        "\n",
        "        # Bind the callback function to the button\n",
        "        submit_button.on_click(on_submit_clicked)\n",
        "\n",
        "        # Add the submit button to the elements to display\n",
        "        elements_to_display.append(submit_button)\n",
        "\n",
        "    input_container.children = elements_to_display\n",
        "\n",
        "    # Set multi_labeling based on the dropdown selection\n",
        "    two_labeling = number_of_inputs == 2  # True for Two-Way False otherwise\n",
        "    three_labeling = number_of_inputs == 3  # True for Three-Way Labeling, False otherwise\n",
        "\n",
        "experiment_type_dropdown.observe(update_input_fields, 'value')\n",
        "\n",
        "# New Output Container for Messages\n",
        "message_output = widgets.Output()\n",
        "\n",
        "# Callback Function for the Button\n",
        "def on_submit_clicked(b):\n",
        "    global sequences\n",
        "    with message_output:\n",
        "        clear_output(wait=True)\n",
        "        sequences = []\n",
        "        entered_sequences = [input_field.value for input_field in text_inputs if input_field.value.strip() != '']\n",
        "        if not entered_sequences:\n",
        "            print(\"\\nPlease enter at least one substrate sequence before pressing the 'Submit' button.\")\n",
        "        else:\n",
        "            sequences = [sequence.replace(\" \", \"\").upper() for sequence in entered_sequences]\n",
        "            if contains_non_ATGC_characters(sequences):\n",
        "                print(\"\\nNon-nucleotide characters detected, please re-enter your sequence(s) and press the 'Submit' button.\")\n",
        "                sequences = []\n",
        "            elif check_long_short_strings(sequences):\n",
        "                print(\"\\nSubstrate sequence(s) of inappropriate length detected, please re-enter your sequence and press the 'Submit' button.\\n\\nSee the instructions above for more details on substrate length restrictions.\")\n",
        "                sequences = []\n",
        "            elif has_duplicates(sequences):\n",
        "                print(\"\\nOne or more of your entered substrate sequences are identical -- please re-enter your sequence and press the 'Submit' button.\\n\\nSee the instructions above for more details on options in orthogonal labeling experiments.\")\n",
        "                sequences = []\n",
        "            else:\n",
        "                print(\"\\nValid Sequence(s) entered - Please proceed to next cell.\")\n",
        "\n",
        "        for input_field in text_inputs:\n",
        "            input_field.value = ''\n",
        "\n",
        "        if sequences:\n",
        "            output = \"\\nEntered Sequence(s):\\n\"\n",
        "            for i, sequence in enumerate(sequences, start=1):\n",
        "                output += f\"Sequence {i}: {sequence}\\n\"\n",
        "            print(output)\n",
        "\n",
        "# Bind the callback function to the button\n",
        "submit_button.on_click(on_submit_clicked)\n",
        "\n",
        "# Display the HTML Label, Dropdown Menu, Input Container, Button, and Message Output\n",
        "display(dropdown_label)\n",
        "display(experiment_type_dropdown)\n",
        "display(input_container)\n",
        "display(message_output)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "NydGKwreA8yy"
      },
      "outputs": [],
      "source": [
        "#@title 3: Select HUH-tag(s) for Labeling\n",
        "\n",
        "#@markdown 1. Run this cell.\n",
        "#@markdown 2. Respond to prompts using the drop-down menus below.\n",
        "#@markdown  * You can either select from the complete list of HUH-tags, or HUHgle can recommend a subset of HUH-tags based on minimizing predicted interaction with the queried substrate sequence(s).\n",
        "#@markdown 3. Proceed to the next cell after making a final selection.\n",
        "\n",
        "###\n",
        "\n",
        "if 'data' not in locals() and 'data' not in globals():\n",
        "    raise NameError(\"PLEASE RUN CELL 1 BEFORE ATTEMPTING TO RUN OTHER CELLS!\")\n",
        "\n",
        "###\n",
        "\n",
        "import ipywidgets as widgets\n",
        "from IPython.display import display, clear_output\n",
        "\n",
        "# Function to make naming in ortho pair dictionaries more clear\n",
        "def convert_frozensets_to_names(frozenset_list):\n",
        "    return [', '.join(nuclease for nuclease in fs) for fs in frozenset_list]\n",
        "\n",
        "def convert_names_to_frozenset(name_string):\n",
        "    nuclease_list = name_string.split(', ')\n",
        "    return frozenset(nuclease_list)\n",
        "\n",
        "def convert_string_to_list(name_string):\n",
        "    return name_string.split(', ')\n",
        "\n",
        "# Function to create a follow-up dropdown with its label\n",
        "def create_followup_dropdown():\n",
        "    followup_label_html = widgets.HTML(\n",
        "        value=\"<b style='min-width: 300px; display: inline-block;'>Select HUH-tag(s):</b>\"\n",
        "    )\n",
        "    followup_dropdown = widgets.Dropdown(\n",
        "        options= nucs,\n",
        "        value=None,\n",
        "        disabled=False\n",
        "    )\n",
        "    followup_dropdown.observe(on_followup_dropdown_change, names='value')\n",
        "    return followup_label_html, followup_dropdown\n",
        "\n",
        "# Function to handle changes in the first dropdown\n",
        "def on_first_dropdown_change(change):\n",
        "    global nucs\n",
        "    nucs = list(dictionaries.keys())\n",
        "    if change['type'] == 'change' and change['name'] == 'value' and change['new']:\n",
        "        if change['new'] == 'Yes, recommend HUH-tag.':\n",
        "            # Follow up for Single - Recommend\n",
        "            nucs = recommend_huh(sequences[0], data)\n",
        "            followup_label, followup_dropdown = create_followup_dropdown()\n",
        "            vbox.children = [first_label_html, first_dropdown, followup_label, followup_dropdown]\n",
        "        elif change['new'] == 'Select my own HUH-tag.':\n",
        "            # Follow up for Single - Select own\n",
        "            followup_label, followup_dropdown = create_followup_dropdown()\n",
        "            vbox.children = [first_label_html, first_dropdown, followup_label, followup_dropdown]\n",
        "        elif change['new'] == 'Yes, recommend HUH-tag pairs.':\n",
        "            # Follow up for Two-Way - Recommend\n",
        "            nucs = recommend_huh_pair(sequences, data, convert_frozensets_to_names(list(top_scoring_pairs.keys())))\n",
        "            followup_label, followup_dropdown = create_followup_dropdown()\n",
        "            vbox.children = [first_label_html, first_dropdown, followup_label, followup_dropdown]\n",
        "        elif change['new'] == 'Select my own HUH-tag pair.':\n",
        "            # Follow up for Two-Way - Select own\n",
        "            nucs = convert_frozensets_to_names(list(top_scoring_pairs.keys()))\n",
        "            followup_label, followup_dropdown = create_followup_dropdown()\n",
        "            vbox.children = [first_label_html, first_dropdown, followup_label, followup_dropdown]\n",
        "        elif change['new'] == 'Yes, recommend HUH-tag triplets.':\n",
        "            # Follow up for Three-Way - Recommend\n",
        "            nucs = recommend_huh_triplet(sequences, data, convert_frozensets_to_names(list(top_scoring_triplets.keys())))\n",
        "            followup_label, followup_dropdown = create_followup_dropdown()\n",
        "            vbox.children = [first_label_html, first_dropdown, followup_label, followup_dropdown]\n",
        "        elif change['new'] == 'Select my own HUH-tag triplet.':\n",
        "            # Follow up for Three-Way - Select own\n",
        "            nucs = convert_frozensets_to_names(list(top_scoring_triplets.keys()))\n",
        "            followup_label, followup_dropdown = create_followup_dropdown()\n",
        "            vbox.children = [first_label_html, first_dropdown, followup_label, followup_dropdown]\n",
        "\n",
        "def on_followup_dropdown_change(change):\n",
        "    global selected_option\n",
        "    if change['type'] == 'change' and change['name'] == 'value' and change['new']:\n",
        "        if two_labeling == False and three_labeling == False:\n",
        "            selected_option = [change['new']]\n",
        "            print(\"\\nHUH-tag(s) Selected - Proceed to Next Cell\")\n",
        "        else:\n",
        "            selected_option = convert_string_to_list(change['new'])\n",
        "            print(\"\\nHUH-tag(s) Selected - Proceed to Next Cell\")\n",
        "\n",
        "# HTML widget for the label of the first dropdown\n",
        "first_label_html = widgets.HTML(\n",
        "    value=\"<b style='min-width: 300px; display: inline-block;'>Recommend HUH-tag(s)?:</b>\"\n",
        ")\n",
        "\n",
        "if len(sequences) == 0:\n",
        "    clear_output(wait=True)\n",
        "    print(\"No Sequence Entered - Please Return to Cell 2\")\n",
        "\n",
        "elif two_labeling == False and three_labeling == False:\n",
        "    first_dropdown = widgets.Dropdown(\n",
        "        options=['', 'Yes, recommend HUH-tag.', 'Select my own HUH-tag.'],\n",
        "        value=None,\n",
        "        disabled=False\n",
        "    )\n",
        "    first_dropdown.observe(on_first_dropdown_change)\n",
        "\n",
        "    # VBox to contain the dropdowns, labels, and button\n",
        "    vbox = widgets.VBox(children=[first_label_html, first_dropdown])\n",
        "\n",
        "    # Display the VBox\n",
        "    display(vbox)\n",
        "\n",
        "elif two_labeling == True:\n",
        "    first_dropdown = widgets.Dropdown(\n",
        "        options=['', 'Yes, recommend HUH-tag pairs.', 'Select my own HUH-tag pair.'],\n",
        "        value=None,\n",
        "        disabled=False\n",
        "    )\n",
        "    first_dropdown.observe(on_first_dropdown_change)\n",
        "\n",
        "    # VBox to contain the dropdowns, labels, and button\n",
        "    vbox = widgets.VBox(children=[first_label_html, first_dropdown])\n",
        "\n",
        "    # Display the VBox\n",
        "    display(vbox)\n",
        "\n",
        "elif three_labeling == True:\n",
        "    first_dropdown = widgets.Dropdown(\n",
        "        options=['', 'Yes, recommend HUH-tag triplets.', 'Select my own HUH-tag triplet.'],\n",
        "        value=None,\n",
        "        disabled=False\n",
        "    )\n",
        "    first_dropdown.observe(on_first_dropdown_change)\n",
        "\n",
        "    # VBox to contain the dropdowns, labels, and button\n",
        "    vbox = widgets.VBox(children=[first_label_html, first_dropdown])\n",
        "\n",
        "    # Display the VBox\n",
        "    display(vbox)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "-sma4kpqCIaj"
      },
      "outputs": [],
      "source": [
        "#@title 4: Search and Plot!\n",
        "#@markdown 1. Run this cell to generate the HUHgle plot.\n",
        "#@markdown  * Make modifications to the sequence by directly interacting with the plot. Cycle through nucleotides via clicking and the plot will dynamically recalculate HUH-tag interaction sites and their cleavage efficiencies.\n",
        "#@markdown  * Press the \"Reset\" button to re-plot your original sequence.\n",
        "#@markdown  * Use the \"Next Substrate\" and \"Next HUH-tag\" buttons (when available) to cycle through your entered substrates or selected HUH-tags, respectively.\n",
        "#@markdown 2. Click on the \"Download Plot & Info\" button to download the plot image and associated product information. This information will also print to the terminal below the plot.\n",
        "\n",
        "###\n",
        "\n",
        "if 'data' not in locals() and 'data' not in globals():\n",
        "    raise NameError(\"PLEASE RUN CELL 1 BEFORE ATTEMPTING TO RUN OTHER CELLS!\")\n",
        "\n",
        "###\n",
        "\n",
        "import ipywidgets as widgets\n",
        "from IPython.display import display, clear_output\n",
        "\n",
        "%matplotlib widget\n",
        "\n",
        "# HUH-tag reactions are generally more efficient if their motif isn't at the\n",
        "# immediate 5' end of a sequence and if there's a second 'C' after the final\n",
        "# 'AC' dinucleotide -- this is commonly found in endogenous contexts\n",
        "\n",
        "# The stuffer gives some padding on the 5' end and the padder adds the 'C' base\n",
        "# to the 3' end to avoid any potential issues\n",
        "\n",
        "stuffer = \"GCTCT\"\n",
        "padder = \"C\"\n",
        "universal_motif = \"TAATATTACC\"\n",
        "\n",
        "if len(sequences) == 0:\n",
        "    print(\"No Sequence(s) Entered -- Please Return to Cell 2\")\n",
        "\n",
        "elif selected_option is None:\n",
        "    print(\"No HUH-tag(s) Selected -- Please Return to Cell 3\")\n",
        "\n",
        "else:\n",
        "\n",
        "    ###\n",
        "\n",
        "    sequence_list = [''] * len(sequences)\n",
        "\n",
        "    cycle_nucs = selected_option\n",
        "\n",
        "    current_HUH_index = 0\n",
        "\n",
        "    current_sequence_index = 0\n",
        "\n",
        "    sequence_intensity_dict = dictionaries[cycle_nucs[current_HUH_index]]\n",
        "\n",
        "    ###\n",
        "\n",
        "    def find_max_kmer_for_column(dataframe, column_name):\n",
        "        max_index = dataframe[column_name].idxmax()\n",
        "        max_kmer = dataframe.loc[max_index, 'kmers']\n",
        "        return max_kmer\n",
        "\n",
        "    def find_optimal_kmer(dataframe, column_names, column_name):\n",
        "        best_score = float('-inf')\n",
        "        best_kmer = None\n",
        "\n",
        "        for index, row in dataframe.iterrows():\n",
        "            current_kmer = row['kmers']\n",
        "            current_reduction = row[column_name]\n",
        "\n",
        "            # Calculate the impact on other nucleases\n",
        "            impact_score = sum(row[name] for name in column_names if name != column_name)\n",
        "\n",
        "            # Scoring strategy (maximize current_reduction and minimize impact)\n",
        "            combined_score = current_reduction - impact_score\n",
        "\n",
        "            # Update best k-mer if a better score is found\n",
        "            if combined_score > best_score:\n",
        "                best_score = combined_score\n",
        "                best_kmer = current_kmer\n",
        "\n",
        "        return best_kmer\n",
        "\n",
        "    # Handle appending HUH-motifs to sequences\n",
        "    if checkbox_state == True:\n",
        "        if len(sequence_list) == 1:\n",
        "            if len(selected_option) == 1:\n",
        "                sequence_list[0] = stuffer + universal_motif + sequences[0] + padder\n",
        "            elif len(selected_option) == 2:\n",
        "                sequence_list[0] = stuffer + find_optimal_kmer(orthogonal_kmers_pairs[frozenset(selected_option)], selected_option, selected_option[0]) + sequences[0] + padder\n",
        "            elif len(selected_option) == 3:\n",
        "                sequence_list[0] = stuffer + find_optimal_kmer(orthogonal_kmers_triplets[frozenset(selected_option)], selected_option, selected_option[0]) + sequences[0] + padder\n",
        "\n",
        "        elif len(sequence_list) == 2:\n",
        "            if len(selected_option) == 2:\n",
        "                sequence_list[0] = stuffer + find_optimal_kmer(orthogonal_kmers_pairs[frozenset(selected_option)], selected_option, selected_option[0]) + sequences[0] + padder\n",
        "                sequence_list[1] = stuffer + find_optimal_kmer(orthogonal_kmers_pairs[frozenset(selected_option)], selected_option, selected_option[1]) + sequences[1] + padder\n",
        "            if len(selected_option) == 3:\n",
        "                sequence_list[0] = stuffer + find_optimal_kmer(orthogonal_kmers_triplets[frozenset(selected_option)], selected_option, selected_option[0]) + sequences[0] + padder\n",
        "                sequence_list[1] = stuffer + find_optimal_kmer(orthogonal_kmers_triplets[frozenset(selected_option)], selected_option, selected_option[1]) + sequences[1] + padder\n",
        "\n",
        "        elif len(sequence_list) == 3:\n",
        "            sequence_list[0] = stuffer + find_optimal_kmer(orthogonal_kmers_triplets[frozenset(selected_option)], selected_option, selected_option[0]) + sequences[0] + padder\n",
        "            sequence_list[1] = stuffer + find_optimal_kmer(orthogonal_kmers_triplets[frozenset(selected_option)], selected_option, selected_option[1]) + sequences[1] + padder\n",
        "            sequence_list[2] = stuffer + find_optimal_kmer(orthogonal_kmers_triplets[frozenset(selected_option)], selected_option, selected_option[2]) + sequences[2] + padder\n",
        "    else:\n",
        "        sequence_list = sequences\n",
        "\n",
        "    ###\n",
        "\n",
        "    button0 = widgets.Button(description=\"Reset\")\n",
        "    button1 = widgets.Button(description=\"Download Plot & Info\")\n",
        "    button2 = widgets.Button(description=\"Next Substrate\")\n",
        "    button3 = widgets.Button(description=\"Next HUH-tag\")\n",
        "\n",
        "\n",
        "    if two_labeling == False and three_labeling == False:\n",
        "        hbox = widgets.HBox([button0, button1])\n",
        "    elif two_labeling == True:\n",
        "        if len(sequence_list) >1:\n",
        "            hbox = widgets.HBox([button0, button1, button2, button3])\n",
        "        else:\n",
        "            hbox = widgets.HBox([button0, button1, button3])\n",
        "    elif three_labeling == True:\n",
        "        if len(sequence_list) >1:\n",
        "            hbox = widgets.HBox([button0, button1, button2, button3])\n",
        "        else:\n",
        "            hbox = widgets.HBox([button0, button1, button3])\n",
        "\n",
        "    display(hbox)\n",
        "\n",
        "    def on_button0_click(b):\n",
        "        with output:\n",
        "            clear_output(wait=True)\n",
        "            plot_dna_sequence(sequence_list, sequence_intensity_dict, show_legend=global_legend)\n",
        "\n",
        "    def on_button1_click(b):\n",
        "        global virus_name, temp\n",
        "        plt.savefig(f'HUHglePlot_{virus_name}_CleavageProfileSubstrate{current_sequence_index + 1}.png', dpi=300)\n",
        "        final_sequence = ''.join(temp)\n",
        "\n",
        "        with open(f'Output_{virus_name}_CleavageProfileSubstrate{current_sequence_index + 1}.txt', 'w') as file:\n",
        "            pass\n",
        "\n",
        "        def custom_print(*args, **kwargs):\n",
        "            # Print to the terminal\n",
        "            print(*args, **kwargs)\n",
        "            # Print to the file\n",
        "            with open(f'Output_{virus_name}_CleavageProfileSubstrate{current_sequence_index + 1}.txt', 'a') as file:\n",
        "                print(*args, **kwargs, file=file)\n",
        "\n",
        "        custom_print(f'Selected Nuclease: {virus_name}\\n')\n",
        "\n",
        "        custom_print(\"Final Modified Sequence:\", final_sequence, \"\\n\")\n",
        "\n",
        "        present_sequences = {sequence: intensity for sequence, intensity in sequence_intensity_dict.items() if sequence in final_sequence}\n",
        "\n",
        "        # Sort the motifs by their first occurrence in the sequence\n",
        "        sorted_motifs = sorted(present_sequences.items(), key=lambda x: final_sequence.find(x[0]))\n",
        "\n",
        "        # Store the start and end locations of each motif\n",
        "        motif_locations = {}\n",
        "\n",
        "        if sorted_motifs:\n",
        "            custom_print(\"Cleaved Motifs in the Modified Sequence:\")\n",
        "            for index, (sequence, intensity) in enumerate(sorted_motifs, start=1):\n",
        "                # Find all locations of the sequence in the final sequence\n",
        "                locations = [range(i+1, i+len(sequence)+1) for i in range(len(final_sequence)) if final_sequence.startswith(sequence, i)]\n",
        "                motif_locations[index] = locations\n",
        "                locations_str = ', '.join([f\"{min(loc)}-{max(loc)}\" for loc in locations])  # Convert list to comma-separated string\n",
        "                custom_print(f\"Motif {index}: {sequence}, Substrate Score: {intensity * 1:.2f}, Location: nucleotides {locations_str}\")\n",
        "        else:\n",
        "            custom_print(\"Cleaved Motifs in the Substrate: None\")\n",
        "\n",
        "        sequences_of_interest = list(sequence_intensity_dict.keys())\n",
        "\n",
        "        primary_products = find_primary_products(final_sequence, sequences_of_interest)\n",
        "\n",
        "        secondary_products = find_secondary_products(primary_products, sequences_of_interest)\n",
        "\n",
        "        # Print primary and secondary products\n",
        "        custom_print(\"\\nPrimary Products of the Modified Sequence:\")\n",
        "        if primary_products:\n",
        "            # Dictionary to associate motifs with their primary products\n",
        "            motif_to_products = {index: [] for index in motif_locations}\n",
        "\n",
        "            for product in primary_products:\n",
        "                # Determine which motif produced this product by checking the endpoints\n",
        "                product_start = final_sequence.find(product) + 1\n",
        "                product_end = product_start + len(product) - 1\n",
        "                product_range = range(product_start, product_end + 1)\n",
        "\n",
        "                for index, locations in motif_locations.items():\n",
        "                    if any(product_range.start in loc or product_range.stop in loc for loc in locations):\n",
        "                        motif_to_products[index].append(product)\n",
        "                        break\n",
        "\n",
        "            # Report products in the order of the motifs\n",
        "            for index in sorted(motif_to_products):\n",
        "                for product in motif_to_products[index]:\n",
        "                    custom_print(f\"From Motif {index}: {virus_name}-{product} (+ {calculate_dna_molecular_weight(product)} kDa)\")\n",
        "\n",
        "        else:\n",
        "            custom_print(\"None\")\n",
        "\n",
        "        custom_print(\"\\nSecondary Products of the Modified Sequence:\")\n",
        "        if secondary_products:\n",
        "            for product in secondary_products:\n",
        "                custom_print(f\"{virus_name}-{product} (+ {calculate_dna_molecular_weight(product)} kDa)\")\n",
        "        else:\n",
        "            custom_print(\"None\")\n",
        "\n",
        "        # Create a ZIP file\n",
        "        with zipfile.ZipFile(f'BatchedFiles_{virus_name}_Substrate{current_sequence_index + 1}.zip', 'w') as zipf:\n",
        "            # Add files\n",
        "            zipf.write(f'HUHglePlot_{virus_name}_CleavageProfileSubstrate{current_sequence_index + 1}.png')\n",
        "            zipf.write(f'Output_{virus_name}_CleavageProfileSubstrate{current_sequence_index + 1}.txt')\n",
        "\n",
        "        # Download the ZIP file\n",
        "        files.download(f'BatchedFiles_{virus_name}_Substrate{current_sequence_index + 1}.zip')\n",
        "\n",
        "        print(\"\\nPlot and Cleavage Information Download Complete\\n\")\n",
        "\n",
        "    def on_button2_click(b):\n",
        "        global current_sequence_index\n",
        "        with output:\n",
        "            clear_output(wait=True)\n",
        "            current_sequence_index = (current_sequence_index + 1) % len(sequence_list)\n",
        "            plot_dna_sequence(sequence_list, sequence_intensity_dict, show_legend=global_legend)\n",
        "\n",
        "    def on_button3_click(b):\n",
        "        global current_HUH_index, sequence_intensity_dict\n",
        "        with output:\n",
        "            clear_output(wait=True)\n",
        "            current_HUH_index = (current_HUH_index + 1) % len(cycle_nucs)\n",
        "            sequence_intensity_dict = dictionaries[cycle_nucs[current_HUH_index]]\n",
        "            plot_dna_sequence(sequence_list, sequence_intensity_dict, show_legend=global_legend)\n",
        "\n",
        "    button0.on_click(on_button0_click)\n",
        "    button1.on_click(on_button1_click)\n",
        "    button2.on_click(on_button2_click)\n",
        "    button3.on_click(on_button3_click)\n",
        "\n",
        "    output = widgets.Output()\n",
        "\n",
        "    with output:\n",
        "        plot_dna_sequence(sequence_list, sequence_intensity_dict, show_legend=global_legend)\n",
        "\n",
        "    display(output)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "JgIgDk0K3fSW"
      },
      "source": [
        "# HUHgle Bonus Features\n",
        "\n",
        "The following cells contain functions that we found to be useful (but not always essential) in the design and optimization of HUH-tag substrates."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "xnu5bLLNPfBj"
      },
      "outputs": [],
      "source": [
        "#@title Bonus Feature #1: 'Expanded' Mode\n",
        "#@markdown 1. Run this cell.\n",
        "#@markdown 2. Toggle the button displayed below on or off to expand the searching capabilities of HUHgle via *in silico* prediction.\n",
        "#@markdown  * Replication-initiating HUH-endonuclease have both sequence- and strucural requirements for ssDNA substrate recognition and interaction. We have observed that the identity of the -4 and +1 positions in a cleavage substrate can be highly variable, but they must be able to form a proper Watson-Crick (WC) base pairing for cleavage to occur (see *anatomy of an HUH-tag reaction* above). Expanded mode broadens the scope of HUHgle by predicting cleavage efficiencies for substrates containing non-cognate WC base pairings between the -4 and +1 positions. Importantly, there are exceptions to this rule and some non-WC combinations can enable cleavage (depending on the HUH-tag).\n",
        "#@markdown 3. **All main and bonus HUHgle functions will use expanded mode when this is switched on.**\n",
        "\n",
        "###\n",
        "\n",
        "if 'data' not in locals() and 'data' not in globals():\n",
        "    raise NameError(\"PLEASE RUN CELL 1 BEFORE ATTEMPTING TO RUN OTHER CELLS!\")\n",
        "\n",
        "###\n",
        "\n",
        "import ipywidgets as widgets\n",
        "from IPython.display import display\n",
        "\n",
        "# Function to expand the dictionaries with new sequences\n",
        "def expand_dictionaries(dicts):\n",
        "    for col, kmers_dict in dicts.items():\n",
        "        new_entries = {}\n",
        "        for kmer, value in kmers_dict.items():\n",
        "            if kmer[3] == \"T\":\n",
        "                # Create new sequences with replacements at index 3 and 7\n",
        "                new_kmers = [\n",
        "                    kmer[:3] + \"A\" + kmer[4:7] + \"T\" + kmer[8:],\n",
        "                    kmer[:3] + \"G\" + kmer[4:7] + \"C\" + kmer[8:],\n",
        "                    kmer[:3] + \"C\" + kmer[4:7] + \"G\" + kmer[8:]\n",
        "                ]\n",
        "                for new_kmer in new_kmers:\n",
        "                    new_entries[new_kmer] = value  # Assign parent sequence value to new sequences\n",
        "        # Update the dictionary with new entries\n",
        "        kmers_dict.update(new_entries)\n",
        "\n",
        "# Create a dictionary for each column\n",
        "dictionaries = {col: create_dict_from_column(data, col) for col in data.columns[2:]}\n",
        "\n",
        "def on_toggle_button_click(change):\n",
        "    global dictionaries\n",
        "    if change['new']:  # Check if the button is pressed\n",
        "        toggle_button.description = 'On'\n",
        "        expand_dictionaries(dictionaries)\n",
        "    else:\n",
        "        toggle_button.description = 'Off'\n",
        "        dictionaries = {col: create_dict_from_column(data, col) for col in data.columns[2:]}\n",
        "\n",
        "toggle_button = widgets.ToggleButton(\n",
        "    value=False,\n",
        "    description='Off',\n",
        "    disabled=False,\n",
        "    button_style='',\n",
        "    tooltip=''\n",
        ")\n",
        "\n",
        "print(\"\\nToggle Expanded Mode:\\n\")\n",
        "\n",
        "# Attach the event handler to the button\n",
        "toggle_button.observe(on_toggle_button_click, names='value')\n",
        "\n",
        "display(toggle_button)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "bzP3oQqDV9mR",
        "cellView": "form"
      },
      "outputs": [],
      "source": [
        "#@title Bonus Feature #2: Cleavage Calculations on Substrates Containing Ambiguous Nucleotides\n",
        "#@markdown 1. Run this cell.\n",
        "#@markdown 2. Enter ssDNA substrate sequence(s) to search (sequences must be between 9 and 300 nt in length).\n",
        "#@markdown  * Please note that you **should not include** an HUH-motif in this kind of analysis because it will complicate interpretation of the results.\n",
        "#@markdown 3. Hit the 'Submit Sequence' button to finalize your substrate submission.\n",
        "#@markdown 4. Select an HUH-tag from the available candidates in the dropdown menu.\n",
        "#@markdown  * Upon selection, substrate analysis will begin and associated files will automatically print to the terminal and download.\n",
        "\n",
        "#@markdown Information regarding restrictions placed on ambiguous nucleotide-containing substrates:\n",
        "#@markdown * Any of the nucleotides outlined in the table below are suitable for use in substrates in this HUHgle bonus feature.\n",
        "#@markdown * We have limited the number of ambiguous nucleotides allowed in a substrate sequence to **eight** nucleotides. Calculations with substrates containing more than eight of these nucleotides are not suitable for the Colab environment.\n",
        "#@markdown    * See HUHgle's GitHub page for access to the source code in order to support a greater number of ambiguous nucleotides, if necessary.\n",
        "\n",
        "#@markdown This table shows the nucleotide ambiguity codes and their corresponding designations:\n",
        "\n",
        "#@markdown | Code | Designation           |\n",
        "#@markdown |------|------------------------|\n",
        "#@markdown | M    | A or C                 |\n",
        "#@markdown | V    | A, C, or G             |\n",
        "#@markdown | R    | A or G                 |\n",
        "#@markdown | H    | A, C, or T             |\n",
        "#@markdown | W    | A or T                 |\n",
        "#@markdown | D    | A, G, or T             |\n",
        "#@markdown | S    | C or G                 |\n",
        "#@markdown | B    | C, G, or T             |\n",
        "#@markdown | Y    | C or T                 |\n",
        "#@markdown | N    | A, C, G, or T          |\n",
        "#@markdown | K    | G or T                 |\n",
        "\n",
        "###\n",
        "\n",
        "if 'data' not in locals() and 'data' not in globals():\n",
        "    raise NameError(\"PLEASE RUN CELL 1 BEFORE ATTEMPTING TO RUN OTHER CELLS!\")\n",
        "\n",
        "###\n",
        "\n",
        "%matplotlib widget\n",
        "\n",
        "from itertools import product\n",
        "import numpy as np\n",
        "import ipywidgets as widgets\n",
        "import matplotlib.pyplot as plt\n",
        "from tqdm import tqdm\n",
        "from google.colab import files\n",
        "import zipfile\n",
        "import os\n",
        "\n",
        "iub_ambiguity_codes = {\n",
        "    \"M\": [\"A\", \"C\"],\n",
        "    \"V\": [\"A\", \"C\", \"G\"],\n",
        "    \"R\": [\"A\", \"G\"],\n",
        "    \"H\": [\"A\", \"C\", \"T\"],\n",
        "    \"W\": [\"A\", \"T\"],\n",
        "    \"D\": [\"A\", \"G\", \"T\"],\n",
        "    \"S\": [\"C\", \"G\"],\n",
        "    \"B\": [\"C\", \"G\", \"T\"],\n",
        "    \"Y\": [\"C\", \"T\"],\n",
        "    \"N\": [\"A\", \"C\", \"G\", \"T\"],\n",
        "    \"K\": [\"G\", \"T\"]\n",
        "}\n",
        "\n",
        "# Global variable definition\n",
        "HUH_choice = None\n",
        "sequence = \"\"\n",
        "\n",
        "###\n",
        "\n",
        "def generate_sequences(sequence):\n",
        "    # Replace each ambiguous nucleotide with the corresponding list from the dictionary\n",
        "    lists_to_product = [iub_ambiguity_codes[nucleotide] if nucleotide in iub_ambiguity_codes else [nucleotide] for nucleotide in sequence]\n",
        "\n",
        "    # Use itertools.product to generate all combinations\n",
        "    return [''.join(seq) for seq in product(*lists_to_product)]\n",
        "\n",
        "def contains_invalid_characters(sequence_list, iub_ambiguity_codes):\n",
        "    # Combine ATGC with the keys from the IUB ambiguity codes dictionary\n",
        "    valid_chars = set(['A', 'T', 'G', 'C']) | set(iub_ambiguity_codes.keys())\n",
        "\n",
        "    # Initialize a dictionary to hold counts for each sequence\n",
        "    ambiguous_char_count = {sequence: 0 for sequence in sequence_list}\n",
        "\n",
        "    for sequence in sequence_list:\n",
        "        for char in sequence:\n",
        "            if char in iub_ambiguity_codes:  # Check if the character is in the IUB codes\n",
        "                ambiguous_char_count[sequence] += 1\n",
        "            elif char not in valid_chars:\n",
        "                print(\"\\nNon-nucleotide characters detected, please re-enter your sequence and press the 'Submit' button.\\n\")\n",
        "                return True\n",
        "\n",
        "    if ambiguous_char_count[sequence] >= 9:\n",
        "        print(\"\\nToo many ambiguous nucleotide characters detected, please re-enter your sequence and press the 'Submit' button.\\n\")\n",
        "        return True\n",
        "\n",
        "    return False\n",
        "\n",
        "def analyze_sequences_by_huh_tag(huh_tag, sequences, huh_tag_data):\n",
        "    if huh_tag not in huh_tag_data:\n",
        "        print(\"HUH-tag data not found.\")\n",
        "        return\n",
        "\n",
        "    # Bins for scores\n",
        "    high_reduction, moderate_reduction, low_reduction = [], [], []\n",
        "    all_reductions = []\n",
        "    all_motifs_and_reductions = []\n",
        "\n",
        "    # Access the specific dictionary for the given HUH-tag\n",
        "    specific_huh_tag_data = huh_tag_data[huh_tag]\n",
        "\n",
        "    for sequence in tqdm(sequences, desc=\"Analyzing sequences\"):\n",
        "        for motif, reduction in specific_huh_tag_data.items():\n",
        "            if motif in sequence:\n",
        "                all_reductions.append(reduction)\n",
        "                all_motifs_and_reductions.append([motif, reduction])\n",
        "                if 1 >= reduction >= 0.75:\n",
        "                    high_reduction.append(reduction)\n",
        "                elif 0.75 > reduction >= 0.35:\n",
        "                    moderate_reduction.append(reduction)\n",
        "                elif 0.35 > reduction >= 0.05:\n",
        "                    low_reduction.append(reduction)\n",
        "\n",
        "    # Prepare output string\n",
        "    print(\"\\n\")\n",
        "    output = (\n",
        "        f\"Total Number of Substrates Analyzed: {len(sequences)}\\n\"\n",
        "        f\"High Scoring Motifs: {len(high_reduction)}\\n\"\n",
        "        f\"Moderate Scoring Motifs: {len(moderate_reduction)}\\n\"\n",
        "        f\"Low Scoring Motifs: {len(low_reduction)}\\n\"\n",
        "        f\"\\nTotal Number of Predicted Cleavage Sites: {len(all_reductions)}\\n\"\n",
        "        f\"\\nKEY:\\nCutoff for High Scoring: 100 - 75%\\n\"\n",
        "        f\"Cutoff for Moderate Scoring: 75 - 35%\\n\"\n",
        "        f\"Cutoff for Low Scoring: 35 - 5%\\n\\n\"\n",
        "    )\n",
        "\n",
        "    # Print the output\n",
        "    print(output)\n",
        "\n",
        "    # Save output to file\n",
        "    text_file = f\"analysis_{huh_tag}.txt\"\n",
        "    with open(text_file, 'w') as file:\n",
        "        file.write(output)\n",
        "\n",
        "    with open(text_file, 'a') as file:\n",
        "        for i in all_motifs_and_reductions:\n",
        "            file.write(f\"Motif: {i[0]}, Substrate Score: {i[1]}\\n\")\n",
        "\n",
        "    # Plotting\n",
        "    plot_file = None\n",
        "    if len(all_reductions) != 0:\n",
        "        plt.figure(figsize=(10, 6))\n",
        "        plt.hist(all_reductions, bins=50, color='#6666FF', edgecolor='None', alpha=0.5)\n",
        "        plt.title(f'Substrate Score Histogram for {huh_tag}', fontsize=28)\n",
        "        plt.xlabel('Substrate Score', fontsize=20)\n",
        "        plt.ylabel('Count', fontsize=20)\n",
        "        plt.xticks(fontsize=15)\n",
        "        plt.yticks(fontsize=15)\n",
        "        plt.xlim(0.0, 1.0)\n",
        "        plt.grid(alpha=0.75)\n",
        "\n",
        "        # Save plot to file\n",
        "        plot_file = f\"histogram_{huh_tag}.png\"\n",
        "        plt.savefig(plot_file)\n",
        "\n",
        "        # Display plot\n",
        "        plt.show()\n",
        "\n",
        "    else:\n",
        "        print(\"No cleavage detected -- plotting canceled.\")\n",
        "\n",
        "    # Create a zip file containing the output text file and plot\n",
        "    zip_file = f\"analysis_{huh_tag}.zip\"\n",
        "    with zipfile.ZipFile(zip_file, 'w') as zipf:\n",
        "        zipf.write(text_file, os.path.basename(text_file))\n",
        "        if plot_file:\n",
        "            zipf.write(plot_file, os.path.basename(plot_file))\n",
        "\n",
        "    # Download the zip file\n",
        "    files.download(zip_file)\n",
        "\n",
        "###\n",
        "\n",
        "print(\"\\nEnter the ssDNA substrate sequence to search ⬇️\\n\")\n",
        "\n",
        "# Text Input Widget for the Sequence\n",
        "single_sequence_input = widgets.Text(description='Sequence:')\n",
        "\n",
        "# HTML Widget for the Dropdown Label\n",
        "dropdown_label_html = widgets.HTML(\n",
        "    value=\"<b>Select HUH-tag:</b>\",\n",
        "    layout=widgets.Layout(margin='0 0 5px 0')\n",
        ")\n",
        "\n",
        "# Dropdown for HUH-tag selection without the built-in description\n",
        "huh_tag_dropdown = widgets.Dropdown(\n",
        "    options=list(dictionaries.keys()),  # Replace with actual options\n",
        "    value=None,\n",
        "    disabled=True\n",
        ")\n",
        "\n",
        "# HTML Widget for the HUH-motif Confirmation Dropdown Label\n",
        "huh_motif_label_html = widgets.HTML(\n",
        "    value=\"<b>Does this sequence contain an intentional HUH-motif that lacks ambiguous characters?</b>\",\n",
        "    layout=widgets.Layout(margin='10px 0 5px 0')\n",
        ")\n",
        "\n",
        "# New Dropdown for HUH-motif Confirmation\n",
        "huh_motif_confirmation_dropdown = widgets.Dropdown(\n",
        "    options=['', 'Yes', 'No'],\n",
        "    description='',\n",
        "    disabled=True\n",
        ")\n",
        "\n",
        "def on_huh_motif_yes_selected():\n",
        "    # slider widget\n",
        "    slider = widgets.IntSlider(\n",
        "        value=0,\n",
        "        min=0,\n",
        "        max=len(sequence[0]),\n",
        "        step=1,\n",
        "        description='',\n",
        "        continuous_update=True\n",
        "    )\n",
        "\n",
        "    # Output widget for displaying the updated sequence\n",
        "    sequence_output = widgets.Output()\n",
        "\n",
        "    # Submit button\n",
        "    submit_button = widgets.Button(description=\"Submit Sequence\")\n",
        "\n",
        "    # Callback function to update the sequence based on the slider value\n",
        "    def update_sequence(change):\n",
        "        global trimmed_sequence\n",
        "        with sequence_output:\n",
        "            sequence_output.clear_output(wait=True)\n",
        "            trimmed_sequence = sequence[0][change['new']:]\n",
        "            print(\"Updated Sequence:\", trimmed_sequence, \"\\n\")\n",
        "\n",
        "    # Callback function for the submit button\n",
        "    def on_submit_clicked(b):\n",
        "        global trimmed_sequence\n",
        "        # Perform the analysis with the modified sequence\n",
        "        print(\"\\n\")\n",
        "        temp = generate_sequences(trimmed_sequence)\n",
        "        analyze_sequences_by_huh_tag(HUH_choice, temp, dictionaries)\n",
        "\n",
        "    # Attach the callback functions to the slider and button\n",
        "    slider.observe(update_sequence, names='value')\n",
        "    submit_button.on_click(on_submit_clicked)\n",
        "\n",
        "    # Initially display the full sequence\n",
        "    with sequence_output:\n",
        "        print(\"Updated Sequence:\", sequence[0], \"\\n\")\n",
        "\n",
        "    # Display the slider and the output widget\n",
        "    display(widgets.VBox([slider, sequence_output, submit_button]))\n",
        "\n",
        "# Callback Function for the HUH-motif Confirmation Dropdown\n",
        "def on_huh_motif_confirmation(change):\n",
        "    if change['type'] == 'change' and change['name'] == 'value':\n",
        "        if change['new'] == 'Yes':\n",
        "            print(\"\\nTrim the HUH-motif away from the rest of your sequence using the slider below:\\n\\n    - Slide right to remove bases from the 5' end of your sequence.\\n\\n    - The presence of HUH-motifs within ambiguous nucleotide-containing sequences (like barcoding oligos) complicates subsequent analyses.\\n\\n    - See the supporting information from the HUHgle manuscript for more details on this.\\n\")\n",
        "            on_huh_motif_yes_selected()  # Call the function to display the slider and output widget\n",
        "            global trimmed_sequence\n",
        "            trimmed_sequence = sequence[0]\n",
        "        elif change['new'] == 'No':\n",
        "            print(\"\\n\")\n",
        "            temp = generate_sequences(sequence[0])\n",
        "            analyze_sequences_by_huh_tag(HUH_choice, temp, dictionaries)\n",
        "\n",
        "# Attach the callback function to the HUH-motif confirmation dropdown\n",
        "huh_motif_confirmation_dropdown.observe(on_huh_motif_confirmation, names='value')\n",
        "\n",
        "# Modified Callback Function for the HUH-tag Dropdown\n",
        "def on_huh_tag_selection(change):\n",
        "    global HUH_choice\n",
        "    if change['type'] == 'change' and change['name'] == 'value':\n",
        "        HUH_choice = change['new']\n",
        "        huh_motif_confirmation_dropdown.disabled = False\n",
        "        display(widgets.VBox([huh_motif_label_html, huh_motif_confirmation_dropdown]))\n",
        "\n",
        "# Attach the modified callback function to the huh_tag_dropdown\n",
        "huh_tag_dropdown.observe(on_huh_tag_selection, names='value')\n",
        "\n",
        "# VBox to arrange the label above the dropdown\n",
        "huh_tag_dropdown_container = widgets.VBox([dropdown_label_html, huh_tag_dropdown])\n",
        "\n",
        "# Button Widget for Submission\n",
        "single_submit_button = widgets.Button(description=\"Submit Sequence\", layout=widgets.Layout(margin='10px 0 0 0'))\n",
        "\n",
        "# New Output Container for Messages\n",
        "single_message_output = widgets.Output()\n",
        "\n",
        "# Callback Function for the Button\n",
        "def on_single_submit_clicked(b):\n",
        "    global sequence\n",
        "    with single_message_output:\n",
        "        clear_output(wait=True)\n",
        "        sequence = [single_sequence_input.value.replace(\" \", \"\").upper()]\n",
        "        if sequence[0] == '':\n",
        "            print(\"\\nPlease enter a substrate sequence before pressing the 'Submit' button.\")\n",
        "        elif check_long_short_strings(sequence):\n",
        "            print(\"\\nSubstrate sequence of inappropriate length detected, please re-enter your sequence and press the 'Submit' button.\\n\\nSee the instructions above for more details on substrate length restrictions.\")\n",
        "        elif contains_invalid_characters(sequence, iub_ambiguity_codes):\n",
        "            print(\"See text above for information regarding constraints placed on sequences containing ambiguous nucleotide characters.\")\n",
        "        else:\n",
        "            print(f\"\\nValid Sequence entered: {sequence[0]}\\n\")\n",
        "            huh_tag_dropdown.disabled = False\n",
        "            display(huh_tag_dropdown_container)\n",
        "\n",
        "        # Clear the text in the input field\n",
        "        single_sequence_input.value = ''  # This line clears the input field\n",
        "\n",
        "# Bind the callback function to the button\n",
        "single_submit_button.on_click(on_single_submit_clicked)\n",
        "\n",
        "# Display the Input Widget, Checkbox Container, Button, and Message Output\n",
        "display(single_sequence_input)\n",
        "display(single_submit_button)\n",
        "display(single_message_output)\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Bonus Feature #3: HUHgle.FASTA\n",
        "#@markdown 1. Run this cell.\n",
        "#@markdown 2. Press the 'Upload File' button to select and upload your .FASTA file containing your DNA sequence of interest.\n",
        "#@markdown    * While substrate sequences can be of any length, they can not contain ambiguous nucleotides.\n",
        "#@markdown 4. Interact with subsequent dropdown menus.\n",
        "#@markdown    * Select an HUH-tag.\n",
        "#@markdown    * Select 'yes' if your file represents a circular sequence so HUHgle can identify sites that span the end-start junction.\n",
        "#@markdown    * Select 'yes' if your file is a viral genome from phylum *Cressdnaviricota* and HUHgle will predict its origin of replication.\n",
        "#@markdown 6. Press the 'submit' button to run analysis.\n",
        "\n",
        "###\n",
        "\n",
        "if 'data' not in locals() and 'data' not in globals():\n",
        "    raise NameError(\"PLEASE RUN CELL 1 BEFORE ATTEMPTING TO RUN OTHER CELLS!\")\n",
        "\n",
        "###\n",
        "\n",
        "print(\"Select and upload a .FASTA file by pressing the 'Choose Files' button below ⬇️\\n\")\n",
        "\n",
        "from google.colab import files\n",
        "import ipywidgets as widgets\n",
        "from IPython.display import display, clear_output\n",
        "import os\n",
        "import RNA\n",
        "import re\n",
        "\n",
        "###\n",
        "\n",
        "def sanitize_filename(filename):\n",
        "    # Remove any characters that are not alphanumeric, underscore, or dot\n",
        "    sanitized_filename = re.sub(r'[^\\w\\-.]', '_', filename)\n",
        "    return sanitized_filename\n",
        "\n",
        "def read_fasta(file_path):\n",
        "    with open(file_path, 'r') as file:\n",
        "        # Initialize variables to hold the ID and sequence\n",
        "        sequence_id = None\n",
        "        sequence_data = []\n",
        "\n",
        "        for line in file:\n",
        "            line = line.strip()\n",
        "            if line.startswith(\">\"):  # Check if it's the ID line\n",
        "                if sequence_id is not None:\n",
        "                    break\n",
        "                sequence_id = line[1:]  # Remove '>' and save the ID\n",
        "                sequence_id = sanitize_filename(sequence_id)  # Sanitize the sequence ID\n",
        "            else:\n",
        "                sequence_data.append(line)  # Add sequence line to the list\n",
        "\n",
        "        return sequence_id, ''.join(sequence_data)  # Join the sequence parts into a single string\n",
        "###\n",
        "\n",
        "# Function to validate the file extension\n",
        "def is_fasta_file(filename):\n",
        "    return filename.lower().endswith('.fasta')\n",
        "\n",
        "# Function to create the dropdown menus and the submit button\n",
        "def create_interaction_elements():\n",
        "    # HTML widgets for labels\n",
        "    label_html_huh_tag = widgets.HTML(\n",
        "        value=\"<b>Select HUH-tag:</b>\",\n",
        "        layout=widgets.Layout(margin='10px 0 5px 0')\n",
        "    )\n",
        "    label_html1 = widgets.HTML(\n",
        "        value=\"<b>Is this a circular sequence?</b>\",\n",
        "        layout=widgets.Layout(margin='10px 0 5px 0')\n",
        "    )\n",
        "    label_html2 = widgets.HTML(\n",
        "        value=\"<b>Is the sequence from phylum Cressdnaviricata?</b>\",\n",
        "        layout=widgets.Layout(margin='10px 0 5px 0')\n",
        "    )\n",
        "\n",
        "    # Dropdown menus\n",
        "    dropdown_menu_huh_tag = widgets.Dropdown(\n",
        "        options=list(dictionaries.keys()),\n",
        "        description='',\n",
        "        disabled=False,\n",
        "    )\n",
        "    dropdown_menu1 = widgets.Dropdown(\n",
        "        options=['No', 'Yes'],\n",
        "        description='',\n",
        "        disabled=False,\n",
        "    )\n",
        "    dropdown_menu2 = widgets.Dropdown(\n",
        "        options=['No', 'Yes'],\n",
        "        description='',\n",
        "        disabled=False,\n",
        "    )\n",
        "\n",
        "    # Button Widget for Submission\n",
        "    submit_button = widgets.Button(description=\"Submit\", layout=widgets.Layout(margin='10px 0 0 0'))\n",
        "\n",
        "    # Display the labels, dropdown menus, and the submit button\n",
        "    display(widgets.VBox([label_html_huh_tag, dropdown_menu_huh_tag, label_html1, dropdown_menu1, label_html2, dropdown_menu2, submit_button]))\n",
        "\n",
        "    # Callback function for the submit button\n",
        "    def on_submit_clicked(b):\n",
        "        clear_output()\n",
        "        print(\"Selections:\")\n",
        "        print(f\"Circular sequence: {'True' if dropdown_menu1.value == 'Yes' else 'False'}\")\n",
        "        print(f\"Sequence from Cressdnaviricota: {'True' if dropdown_menu2.value == 'Yes' else 'False'}\")\n",
        "\n",
        "        # Check if the sequence is circular and adjust accordingly\n",
        "        if dropdown_menu1.value == 'Yes':\n",
        "            sequence_to_search = fasta_file[1][-30:] + fasta_file[1]  # Add the last 30 bp to the front\n",
        "            circular = True\n",
        "        else:\n",
        "            sequence_to_search = fasta_file[1]  # Use the sequence as is for linear DNA\n",
        "            circular = False\n",
        "\n",
        "        if dropdown_menu2.value == 'Yes':\n",
        "            cress_bool = True\n",
        "        else:\n",
        "            cress_bool = False\n",
        "\n",
        "        if contains_non_ATGC_characters(fasta_file[1]) is True:\n",
        "            print(\"\\nNon-nucleotide characters detected, please re-run this cell and upload a .FASTA file that contains only As, Ts, Gs, or Cs.\")\n",
        "\n",
        "        else:\n",
        "            sequence_intensity_dict = dictionaries[dropdown_menu_huh_tag.value]\n",
        "\n",
        "            with open(f'Output_{dropdown_menu_huh_tag.value}_CleavageProfileSubstrate_GB:{fasta_file[0]}.txt', 'w') as file:\n",
        "                pass\n",
        "\n",
        "            def custom_print(*args, **kwargs):\n",
        "                # Print to the file\n",
        "                with open(f'Output_{dropdown_menu_huh_tag.value}_CleavageProfileSubstrate_GB:{fasta_file[0]}.txt', 'a') as file:\n",
        "                    print(*args, **kwargs, file=file)\n",
        "\n",
        "            custom_print(f'Selected Nuclease: {dropdown_menu_huh_tag.value}\\n')\n",
        "\n",
        "            present_sequences = {sequence: intensity for sequence, intensity in sequence_intensity_dict.items() if sequence in sequence_to_search}\n",
        "\n",
        "            # Sort the motifs by their first occurrence in the sequence\n",
        "            sorted_motifs = sorted(present_sequences.items(), key=lambda x: sequence_to_search.find(x[0]))\n",
        "\n",
        "            # Adjust locations for motifs found in the appended section for circular sequences\n",
        "            adjusted_motif_locations = []\n",
        "            fold_motif_information = []\n",
        "\n",
        "            for motif, intensity in sorted_motifs:\n",
        "                motif_locations = [i for i in range(len(sequence_to_search)) if sequence_to_search.startswith(motif, i)]\n",
        "\n",
        "                for location in motif_locations:\n",
        "                    if circular and location < 30:\n",
        "                        # Motif starts within the appended section, adjust position to reflect original sequence\n",
        "                        adjusted_location = [len(fasta_file[1]) - 30 + location, location - 30 + len(motif)]\n",
        "                        adjusted_motif_locations.append([adjusted_location, motif, intensity])\n",
        "                        fold_motif_information.append([[location, location + len(motif)], motif, intensity])\n",
        "                    elif circular and location >= 30:\n",
        "                        adjusted_location = [location - 30, location - 30 + len(motif)]\n",
        "                        adjusted_motif_locations.append([adjusted_location, motif, intensity])\n",
        "                        fold_motif_information.append([[location, location + len(motif)], motif, intensity])\n",
        "                    elif not circular:\n",
        "                        adjusted_location = [location, location + len(motif)]\n",
        "                        adjusted_motif_locations.append([adjusted_location, motif, intensity])\n",
        "                        fold_motif_information.append([[location, location + len(motif)], motif, intensity])\n",
        "\n",
        "            if len(adjusted_motif_locations) >= 1:\n",
        "                custom_print(\"Cleaved Motifs in the Sequence of Interest:\")\n",
        "                for i in range(len(adjusted_motif_locations)):\n",
        "                    custom_print(f\"Motif {i + 1}: {adjusted_motif_locations[i][1]}, Substrate Score: {adjusted_motif_locations[i][2] * 1: .2f}, Location: nucleotides {adjusted_motif_locations[i][0][0]}-{adjusted_motif_locations[i][0][1]}\")\n",
        "            else:\n",
        "                custom_print(\"Cleaved Motifs in the Sequence of Interest: None\")\n",
        "\n",
        "            if cress_bool:\n",
        "                fold_motif_information = sorted(fold_motif_information, key=lambda x: x[2], reverse=True)\n",
        "                fold_motif_information = fold_motif_information[:4]\n",
        "\n",
        "                for i in fold_motif_information:\n",
        "                    sequence_to_fold = sequence_to_search[i[0][0]-10:i[0][1]+10]\n",
        "                    ss, mfe = RNA.fold(sequence_to_fold)\n",
        "                    i.append(sequence_to_fold)\n",
        "                    i.append(ss)\n",
        "                    i.append(mfe)\n",
        "\n",
        "                fold_motif_information = sorted(fold_motif_information,\n",
        "                                                  key=lambda x: x[5])\n",
        "\n",
        "                likely_candidate = fold_motif_information[0]\n",
        "\n",
        "                fold_file = f\"StemLoop{dropdown_menu_huh_tag.value}_{fasta_file[0]}.svg\"\n",
        "\n",
        "                RNA.svg_rna_plot(likely_candidate[3], likely_candidate[4], fold_file)\n",
        "\n",
        "                out_statement = f\"\\nPredicted Origin of Replication: {likely_candidate[1]}\"\n",
        "\n",
        "                custom_print(out_statement)\n",
        "                print(out_statement)\n",
        "\n",
        "            # Plotting\n",
        "            plot_file = None\n",
        "            substrate_scores = [score for sequence, score in sorted_motifs]\n",
        "            if len(substrate_scores) != 0:\n",
        "                plt.figure(figsize=(10, 6))\n",
        "                plt.hist(substrate_scores, bins=50, color='#6666FF', edgecolor='None', alpha=0.5)\n",
        "                plt.title(f'Substrate Score Histogram for {dropdown_menu_huh_tag.value}', fontsize=28)\n",
        "                plt.xlabel('Substrate Score', fontsize=20)\n",
        "                plt.ylabel('Count', fontsize=20)\n",
        "                plt.xticks(fontsize=15)\n",
        "                plt.yticks(fontsize=15)\n",
        "                plt.xlim(0.0, 1.0)\n",
        "                plt.grid(alpha=0.75)\n",
        "\n",
        "                # Save plot to file\n",
        "                plot_file = f\"histogram_{dropdown_menu_huh_tag.value}_GB:{fasta_file[0]}.png\"\n",
        "                plt.savefig(plot_file)\n",
        "\n",
        "                print(\"\\nAnalysis complete. Downloading files.\\n\")\n",
        "\n",
        "                # Display plot\n",
        "                plt.show()\n",
        "\n",
        "            out_text_str = (f'Output_{dropdown_menu_huh_tag.value}_CleavageProfileSubstrate_GB:{fasta_file[0]}.txt')\n",
        "\n",
        "            if cress_bool:\n",
        "                out_list = [plot_file, out_text_str, fold_file]\n",
        "            else:\n",
        "                out_list = [plot_file, out_text_str]\n",
        "\n",
        "            master_zip_file = \"HUHgleFASTAFiles.zip\"\n",
        "            with zipfile.ZipFile(master_zip_file, 'w') as master_zip:\n",
        "                for link in out_list:\n",
        "                    master_zip.write(link)\n",
        "\n",
        "            files.download(master_zip_file)\n",
        "\n",
        "    # Bind the callback function to the button\n",
        "    submit_button.on_click(on_submit_clicked)\n",
        "\n",
        "# Function to upload files and create dropdown menus and the submit button upon successful upload\n",
        "def upload_files_and_create_interaction_elements():\n",
        "    uploaded_files = files.upload()\n",
        "\n",
        "    # Check if files were uploaded and validate file extension\n",
        "    if uploaded_files:\n",
        "        for filename in uploaded_files.keys():\n",
        "            if is_fasta_file(filename):\n",
        "                print(f\"\\nUploaded file '{filename}' is a valid FASTA file.\")\n",
        "                create_interaction_elements()\n",
        "                if uploaded_files:\n",
        "                    # Process each uploaded file\n",
        "                    for filename, content in uploaded_files.items():\n",
        "                        return(read_fasta(filename))\n",
        "            else:\n",
        "                print(f\"\\nUploaded file '{filename}' is not a FASTA file. Please upload a .FASTA file.\")\n",
        "\n",
        "# Main execution\n",
        "fasta_file = upload_files_and_create_interaction_elements()\n"
      ],
      "metadata": {
        "id": "slLx8uNyjIwg",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}